### ISSUE TITLE
Additional Class / File for Lab Assignments and PE

### ISSUE THREAD

**Issue by cliftonfelix** [nus-cs2030/2122-s1#45]
For Lab assignments and PE, can we create extra classes that can help our code to run aside from all the stated classes in the problem statement?

I tried creating extra classes for lab 2 (which must be separated into several files - 1 class in each file) and when I submit to CodeCrunch, there is an error because the test cases do not open the additional files that I created. So, the class is not declared.

**Comment by RussellDash332** [comment]
Probably the Codecrunch tester tries to run only specific files that match and compares the output afterward. So, you might want to restrict your filenames to the ones required in the prompt.

For Lab2, Codecrunch doesn't actually need the `.jsh` test files on your submission because they will run it behind the screen as the autograder. So, having it in your submission is just pure redundancy. However, the reason why it's there is because it was automatically submitted from your PE Nodes for you to download it again back at home.

**Comment by Enigmatrix** [comment]
`~/.backup` should be a directory, not a file. To create such a directory, you can do `mkdir ~/.backup`. To make sure it is accessible for writing (it should be, unless something strange has occurred), you can change the permissions via `chmod 644 ~/.backup`

**Comment by shleen** [comment]
Just out of curiosity, I'm wondering what the reasoning is behind the restriction for CodeCrunch to only include the filenames required in the prompt?

It seems to me that restricting this forces everyone into one "correct" design for their classes, when in fact there might be multiple ways to solve the same problem - each with its own strengths & weaknesses. I'd think that giving us the space to explore & experiment with different designs will help us learn a lot more quickly & deeply why a certain design might be better than another.

I can't imagine it being a technical difficulty - including `*.java` files seems trivial enough. Maybe for security reasons then? Though that doesn't seem like a completely convincing reason to restrict our submitted filenames so heavily.

**Comment by RussellDash332** [comment]
Turns out for the Project, there is no restriction on your class names except for the Simulator class and the driver classes. So, I guess it depends on how are you calling the other classes correctly.

**Comment by shleen** [comment]
That makes sense! I guess my question was more directed to the labs rather than the project.

**Comment by RussellDash332** [comment]
But now we've found a general conclusion, good catch also 😉👍

**Comment by shleen** [comment]
Hm not sure if I'm missing something, but I wouldn't be able to edit this `test9.jsh` file in your example to `/open` my additional files right? Like, I'd be able to do it locally where I can edit `test9.jsh` but doesn't CodeCrunch use its own copies of the test files (that I cannot edit)?

**Comment by RussellDash332** [comment]
Yes, PE nodes are giving u the test jsh files for your reference but whether you submit or modify them or not at all, Code Crunch will use their own test jsh files and overwrites your test jsh submission files.
